docs/proposals/enable-local-source-code-support.md

```diff
@@ -0,0 +1,302 @@
+
+### User Stories [optional]
+
+Build users need to define the required parameter values if they want to opt-in for the usage of local source code feature.
+
+#### As a Shipwright/Build contributor I want to have a well defined API to support local source code
+
+The Build resource API needs to provide means of signalizing the desire of local source code. In terms of API changes this should be minimal.
+
+Therefore, users should be able to signalize this feature via:
+
+```yaml
+spec:
+  source:
+    container:
+      image: docker.io/<registry-repository>/bundle:latest
+      path: /<path-to-local>/github.com/shipwright-io/sample-go/docker-build
+```
+
+#### As a Shipwright User I want to build images in Development mode from my local Dockerfile or plain source code
+
+Users should only need to specify the mandatory fields under `spec.source.container` without modifying their existing strategies of choice. Local source code feature should work out of the box with different tools, like Kaniko, BuildKit, Buildpacks, et al.
+
+```yaml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: Build
+metadata:
+  name: a-kaniko-build
+spec:
+  source:
+    container:
+      image: docker.io/<registry-repository>/bundle:latest
+      local: /<path-to-local>/github.com/shipwright-io/sample-go/docker-build
+  strategy:
+    name: kaniko # Note that users could also specify buildpacks.
+    kind: ClusterBuildStrategy
+  output: ...
+```
+
+#### As a Shipwright User I want to ensure reproducible bundle images
+
+We want to provide trust to users when building bundle images with their source code, so that on multiple builds with the same local source code, the bundle image always express the same _SHA_. Subsequent Builds should generate the same image if the source code is unchanged.
+
+#### As a Shipwright User I want to avoid excessive registry storage from Bundle images
+
+We want to allow users to signalize the desire of pruning a bundle image that was used for building their application container image.
+
+Adding `pruneBundle` should ensure a mechanism for deleting `docker.io/<registry-repository>/bundle:latest`.
+
+```yaml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: Build
+metadata:
+  name: a-kaniko-build
+spec:
+  source:
+    container:
+      image: docker.io/<registry-repository>/bundle:latest
+      local: /<path-to-local>/github.com/shipwright-io/sample-go/docker-build
+      pruneBundle: true # Disable by default
+  strategy:
+    name: kaniko
+    kind: ClusterBuildStrategy
+  output: ...
+```
+
+The above Build definition should ensure that a step in our [TaskSpec](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/reconciler/buildrun/resources/taskrun.go#L59) definition is appended, which will ensure a pruning of the image in the related container registry.
```

Comment:
I like the feature. Though, I think it is complicated to do this across the different container registry providers (see Matthias' `build-load` tool that also deletes images). A fallback could be that we instead only remove all layers that are not in the base image from the image (which effectively means to push the base image).

Theoretically, we also do not need an additional step. The container extraction step could also do this. Though, that would mean that the logic inside that contains becomes more complex.

Reply:
Yes. For now I will add a _call out_ for risk in here.